src/update_agent/mod.rs

```diff
@@ -198,9 +228,86 @@ impl UpdateAgentState {
         self.transition_to(target);
     }
 
-    /// Transition to the UpdateStaged state.
+    /// Transition to the UpdateStaged state, setting the number of postponements
+    /// remaining to `MAX_FINALIZE_POSTPONEMENTS`.
     fn update_staged(&mut self, update: Release) {
-        let target = UpdateAgentState::UpdateStaged(update);
+        let target = UpdateAgentState::UpdateStaged((update, MAX_FINALIZE_POSTPONEMENTS));
+
+        self.transition_to(target);
+    }
+
+    /// Determine whether to allow finalization based off of current state.
+    /// Returns a boolean indicating whether a finalization is permitted.
+    fn usersessions_can_finalize(&mut self) -> bool {
+        match get_interactive_user_sessions() {
+            Ok(interactive_sessions) => {
+                DETECTED_ACTIVE_USERS.set(interactive_sessions.len().try_into().unwrap());
+                self.handle_interactive_sessions(&interactive_sessions)
+            }
+            Err(e) => {
+                // If we failed to check for interactive sessions, just allow
+                // finalization.
+                log::error!("failed to check for interactive sessions: {}", e);
+                true
+            }
+        }
+    }
+
+    /// Helper for determining whether to allow a finalization by first checking whether
+    /// interactive sessions are present and then handling the appropriate response to current
+    /// state's remaining postponements (possibly broadcasting warning messages to active sessions).
+    ///
+    /// Returns a boolean indicating whether finalization is permitted.
+    fn handle_interactive_sessions(&mut self, interactive_sessions: &[InteractiveSession]) -> bool {
+        if interactive_sessions.is_empty() {
+            return true;
+        }
+
+        let (release, postponements_remaining) = match self.clone() {
+            UpdateAgentState::UpdateStaged((r, p)) => (r, p),
+            _ => unreachable!(
```

Comment:
Self-note: later on we should probably try to re-arrange the internals of the FSM data-structure so that it doesn't require this matching.